File: Samples/GSMSample_4_0_Mango_VB/GameStateManagementSample/GameStateManagementSample/Screens/GameplayScreen.vb
#Region "File Description"
'-----------------------------------------------------------------------------
' GameplayScreen.vb
'
' Microsoft XNA Community Game Platform
' Copyright (C) Microsoft Corporation. All rights reserved.
'-----------------------------------------------------------------------------
#End Region

#Region "Using Statements"
Imports Microsoft.VisualBasic
Imports System.Threading
Imports Microsoft.Xna.Framework
Imports Microsoft.Xna.Framework.Content
Imports Microsoft.Xna.Framework.Graphics
Imports Microsoft.Xna.Framework.Input
Imports GameStateManagement
#End Region

''' <summary>
''' This screen implements the actual game logic. It is just a
''' placeholder to get the idea across: you'll probably want to
''' put some more interesting gameplay in here!
''' </summary>
Friend Class GameplayScreen
    Inherits GameScreen
#Region "Fields"

    Private content As ContentManager
    Private gameFont As SpriteFont

    Private playerPosition As New Vector2(100, 100)
    Private enemyPosition As New Vector2(100, 100)

    Private random As New Random()

    Private pauseAlpha As Single

    Private pauseAction As InputAction

#End Region

#Region "Initialization"


    ''' <summary>
    ''' Constructor.
    ''' </summary>
    Public Sub New()
        TransitionOnTime = TimeSpan.FromSeconds(1.5)
        TransitionOffTime = TimeSpan.FromSeconds(0.5)

        pauseAction = New InputAction(New Buttons() {Buttons.Start, Buttons.Back}, New Keys() {Keys.Escape}, True)
    End Sub


    ''' <summary>
    ''' Load graphics content for the game.
    ''' </summary>
    Public Overrides Sub Activate(ByVal instancePreserved As Boolean)
        If Not instancePreserved Then
            If content Is Nothing Then
                content = New ContentManager(ScreenManager.Game.Services, "Content")
            End If

            gameFont = content.Load(Of SpriteFont)("gamefont")

            ' A real game would probably have more content than this sample, so
            ' it would take longer to load. We simulate that by delaying for a
            ' while, giving you a chance to admire the beautiful loading screen.
            Thread.Sleep(1000)

            ' once the load has finished, we use ResetElapsedTime to tell the game's
            ' timing mechanism that we have just finished a very long frame, and that
            ' it should not try to catch up.
            ScreenManager.Game.ResetElapsedTime()
        End If

#If WINDOWS_PHONE Then
        If Microsoft.Phone.Shell.PhoneApplicationService.Current.State.ContainsKey("PlayerPosition") Then
            playerPosition = CType(Microsoft.Phone.Shell.PhoneApplicationService.Current.State("PlayerPosition"), Vector2)
            enemyPosition = CType(Microsoft.Phone.Shell.PhoneApplicationService.Current.State("EnemyPosition"), Vector2)
        End If
#End If
    End Sub


    Public Overrides Sub Deactivate()
#If WINDOWS_PHONE Then
        Microsoft.Phone.Shell.PhoneApplicationService.Current.State("PlayerPosition") = playerPosition
        Microsoft.Phone.Shell.PhoneApplicationService.Current.State("EnemyPosition") = enemyPosition
#End If

        MyBase.Deactivate()
    End Sub


    ''' <summary>
    ''' Unload graphics content used by the game.
    ''' </summary>
    Public Overrides Sub Unload()
        content.Unload()

#If WINDOWS_PHONE Then
        Microsoft.Phone.Shell.PhoneApplicationService.Current.State.Remove("PlayerPosition")
        Microsoft.Phone.Shell.PhoneApplicationService.Current.State.Remove("EnemyPosition")
#End If
    End Sub


#End Region

#Region "Update and Draw"


    ''' <summary>
    ''' Updates the state of the game. This method checks the GameScreen.IsActive
    ''' property, so the game will stop updating when the pause menu is active,
    ''' or if you tab away to a different application.
    ''' </summary>
    Public Overrides Sub Update(ByVal gameTime As GameTime, ByVal otherScreenHasFocus As Boolean, ByVal coveredByOtherScreen As Boolean)
        MyBase.Update(gameTime, otherScreenHasFocus, False)

        ' Gradually fade in or out depending on whether we are covered by the pause screen.
        If coveredByOtherScreen Then
            pauseAlpha = Math.Min(pauseAlpha + 1.0F / 32, 1)
        Else
            pauseAlpha = Math.Max(pauseAlpha - 1.0F / 32, 0)
        End If

        If IsActive Then
            ' Apply some random jitter to make the enemy move around.
            Const randomization As Single = 10

            enemyPosition.X += CSng(random.NextDouble() - 0.5) * randomization
            enemyPosition.Y += CSng(random.NextDouble() - 0.5) * randomization

            ' Apply a stabilizing force to stop the enemy moving off the screen.
            Dim targetPosition As New Vector2(ScreenManager.GraphicsDevice.Viewport.Width \ 2 - gameFont.MeasureString("Insert Gameplay Here").X / 2, 200)

            enemyPosition = Vector2.Lerp(enemyPosition, targetPosition, 0.05F)

            ' TODO: this game isn't very fun! You could probably improve
            ' it by inserting something more interesting in this space :-)
        End If
    End Sub


    ''' <summary>
    ''' Lets the game respond to player input. Unlike the Update method,
    ''' this will only be called when the gameplay screen is active.
    ''' </summary>
    Public Overrides Sub HandleInput(ByVal gameTime As GameTime, ByVal input As InputState)
        If input Is Nothing Then
            Throw New ArgumentNullException("input")
        End If

        ' Look up inputs for the active player profile.
        Dim playerIndex As Integer = CInt(ControllingPlayer.Value)

        Dim keyboardState As KeyboardState = input.CurrentKeyboardStates(playerIndex)
        Dim gamePadState As GamePadState = input.CurrentGamePadStates(playerIndex)

        ' The game pauses either if the user presses the pause button, or if
        ' they unplug the active gamepad. This requires us to keep track of
        ' whether a gamepad was ever plugged in, because we don't want to pause
        ' on PC if they are playing with a keyboard and have no gamepad at all!
        Dim gamePadDisconnected As Boolean = (Not gamePadState.IsConnected) AndAlso input.GamePadWasConnected(playerIndex)

        Dim player As PlayerIndex
        If pauseAction.Evaluate(input, ControllingPlayer, player) OrElse gamePadDisconnected Then
#If WINDOWS_PHONE Then
            ScreenManager.AddScreen(New PhonePauseScreen(), ControllingPlayer)
#Else
				ScreenManager.AddScreen(New PauseMenuScreen(), ControllingPlayer)
#End If
        Else
            ' Otherwise move the player position.
            Dim movement As Vector2 = Vector2.Zero

            If keyboardState.IsKeyDown(Keys.Left) Then
                movement.X -= 1
            End If

            If keyboardState.IsKeyDown(Keys.Right) Then
                movement.X += 1
            End If

            If keyboardState.IsKeyDown(Keys.Up) Then
                movement.Y -= 1
            End If

            If keyboardState.IsKeyDown(Keys.Down) Then
                movement.Y += 1
            End If

            Dim thumbstick As Vector2 = gamePadState.ThumbSticks.Left

            movement.X += thumbstick.X
            movement.Y -= thumbstick.Y

            If input.TouchState.Count > 0 Then
                Dim touchPosition As Vector2 = input.TouchState(0).Position
                Dim direction As Vector2 = touchPosition - playerPosition
                direction.Normalize()
                movement += direction
            End If

            If movement.Length() > 1 Then
                movement.Normalize()
            End If

            playerPosition += movement * 8.0F
        End If
    End Sub


    ''' <summary>
    ''' Draws the gameplay screen.
    ''' </summary>
    Public Overrides Sub Draw(ByVal gameTime As GameTime)
        ' This game has a blue background. Why? Because!
        ScreenManager.GraphicsDevice.Clear(ClearOptions.Target, Color.CornflowerBlue, 0, 0)

        ' Our player and enemy are both actually just text strings.
        Dim spriteBatch As SpriteBatch = ScreenManager.SpriteBatch

        spriteBatch.Begin()

        spriteBatch.DrawString(gameFont, "// TODO", playerPosition, Color.Green)

        spriteBatch.DrawString(gameFont, "Insert Gameplay Here", enemyPosition, Color.DarkRed)

        spriteBatch.End()

        ' If the game is transitioning on or off, fade it out to black.
        If TransitionPosition > 0 OrElse pauseAlpha > 0 Then
            Dim alpha As Single = MathHelper.Lerp(1.0F - TransitionAlpha, 1.0F, pauseAlpha / 2)

            ScreenManager.FadeBackBufferToBlack(alpha)
        End If
    End Sub


#End Region
End Class

File: Samples/Minjie_ARCHIVE_2_0/Minjie/Minjie/Screens/GameplayScreen.cs
#region File Description
//-----------------------------------------------------------------------------
// GameplayScreen.cs
//
// Microsoft XNA Community Game Platform
// Copyright (C) Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------
#endregion

#region Using Statements
using System;
using System.Threading;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Content;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;
#endregion

namespace Minjie
{
    /// <summary>
    /// This screen implements the actual game logic.
    /// </summary>
    /// <remarks>Based on a similar class in the Game State Management sample.</remarks>
    class GameplayScreen : GameScreen
    {
        #region Constant Data


        /// <summary>
        /// The size of the game board.
        /// </summary>
        const int boardSize = 10;

        /// <summary>
        /// The size of the tile model.
        /// </summary>
        const float tileSize = 40.2f;

        /// <summary>
        /// The starting height of the pieces over the board.
        /// </summary>
        const float startingPieceHeight = 50f;

        /// <summary>
        /// The number of frames in the falling-piece animation.
        /// </summary>
        const int fallingPieceFrames = 10;


        #endregion
        
        #region Fields


        ContentManager content;
        Texture2D backgroundTexture;
        Model tileModel, highlightTileModel;
        Model whitePieceModel, blackPieceModel;
        Model whitePieceTileModel, blackPieceTileModel;
        Vector3 boardPosition;
        int fallingPieceFrame = 0;
        int lastPieceCount = 0;

        Board board = new Board(boardSize);
        Player player1, player2;


        #endregion

        #region Initialization


        /// <summary>
        /// Constructor.
        /// </summary>
        public GameplayScreen(bool multiplayer)
        {
            board.Initialize();
            lastPieceCount = board.PieceCount;

            player1 = new LocalPlayer(BoardColors.White, PlayerIndex.One, boardSize);
            if (multiplayer)
            {
                player2 = new LocalPlayer(BoardColors.Black, PlayerIndex.Two,
                    boardSize);
            }
            else
            {
                player2 = new AiPlayer(BoardColors.Black);
            }

            float boardPositionValue = -0.5f * tileSize * (float)boardSize + 
                tileSize / 2f;
            boardPosition = new Vector3(boardPositionValue, 1f, boardPositionValue);

            AudioManager.PlayMusic("Music_Game");
        }


        /// <summary>
        /// Load graphics content for the game.
        /// </summary>
        public override void LoadContent()
        {
            if (content == null)
                content = new ContentManager(ScreenManager.Game.Services, "Content");

            backgroundTexture = content.Load<Texture2D>("Gameplay/game_screen");
            tileModel = content.Load<Model>("Gameplay//tile");
            highlightTileModel = content.Load<Model>("Gameplay//tile_highlight");
            whitePieceModel = content.Load<Model>("Gameplay//p2_piece");
            blackPieceModel = content.Load<Model>("Gameplay//p1_piece");
            whitePieceTileModel = content.Load<Model>("Gameplay//p2_piece_tile");
            blackPieceTileModel = content.Load<Model>("Gameplay//p1_piece_tile");

            // reset the camera
            Viewport viewport = ScreenManager.GraphicsDevice.Viewport;
            SimpleArcCamera.Reset((float)viewport.Width / (float)viewport.Height);

            // once the load has finished, we use ResetElapsedTime to tell the game's
            // timing mechanism that we have just finished a very long frame, and that
            // it should not try to catch up.
            ScreenManager.Game.ResetElapsedTime();
        }


        /// <summary>
        /// Unload graphics content used by the game.
        /// </summary>
        public override void UnloadContent()
        {
            content.Unload();
        }


        #endregion

        #region Update and Draw


        /// <summary>
        /// Updates the state of the game. This method checks the GameScreen.IsActive
        /// property, so the game will stop updating when the pause menu is active,
        /// or if you tab away to a different application.
        /// </summary>
        public override void Update(GameTime gameTime, bool otherScreenHasFocus,
                                                       bool coveredByOtherScreen)
        {
            base.Update(gameTime, otherScreenHasFocus, coveredByOtherScreen);

            if (IsActive)
            {
                SimpleArcCamera.Update(gameTime);
                if (fallingPieceFrame > 0)
                {
                    fallingPieceFrame--;
                    if (fallingPieceFrame <= 0)
                    {
                        AudioManager.PlayCue("Drop");
                    }
                }
                if (fallingPieceFrame <= 0)
                {
                    player1.Update(board);
                    if (board.PieceCount > lastPieceCount)
                    {
                        lastPieceCount = board.PieceCount;
                        fallingPieceFrame = fallingPieceFrames;
                    }
                }
                if (fallingPieceFrame <= 0)
                {
                    player2.Update(board);
                    if (board.PieceCount > lastPieceCount)
                    {
                        lastPieceCount = board.PieceCount;
                        fallingPieceFrame = fallingPieceFrames;
                    }
                }
                if (board.GameOver)
                {
                    GameResult gameResult = GameResult.Tied;
                    if (board.WhitePieceCount > board.BlackPieceCount)
                    {
                        gameResult = GameResult.Player1Won;
                    }
                    else if (board.BlackPieceCount > board.WhitePieceCount)
                    {
                        gameResult = GameResult.Player2Won;
                    }
                    ExitScreen();
                    ScreenManager.AddScreen(new GameOverScreen(gameResult));
                }
            }
        }


        /// <summary>
        /// Lets the game respond to player input. Unlike the Update method,
        /// this will only be called when the gameplay screen is active.
        /// </summary>
        public override void HandleInput(InputState input)
        {
            if (input == null)
                throw new ArgumentNullException("input");

            if (input.ExitGame)
            {
                ExitScreen();
                ScreenManager.AddScreen(new TitleScreen());
            }

            // update the input state for local players and the camera
            SimpleArcCamera.InputState = input;
            LocalPlayer.InputState = input;
        }


        /// <summary>
        /// Draws the gameplay screen.
        /// </summary>
        public override void Draw(GameTime gameTime)
        {
            // This game has a blue background. Why? Because!
            ScreenManager.GraphicsDevice.Clear(Color.White);

            // Our player and enemy are both actually just text strings.
            SpriteBatch spriteBatch = ScreenManager.SpriteBatch;

            spriteBatch.Begin();
            spriteBatch.Draw(backgroundTexture, Vector2.Zero, Color.White);
            spriteBatch.End();

            ScreenManager.GraphicsDevice.RenderState.DepthBufferEnable = true;
            Point cursorPosition = LocalPlayer.CursorPosition;
            for (int row = 0; row < boardSize; row++)
            {
                for (int column = 0; column < boardSize; column++)
                {
                    if ((fallingPieceFrame > 0) && (board.LastMove.Row == row) &&
                        (board.LastMove.Column == column))
                    {
                        int fall = fallingPieceFrames - fallingPieceFrame;
                        float height = startingPieceHeight - (fall + fall * fall / 2);
                        // do not let the piece fall through the board
                        if (height < 6)
                        {
                            height = 6;
                        }
                        DrawModel(board.CurrentColor == BoardColors.White ?
                            blackPieceModel : whitePieceModel, row, column, height);
                        DrawModel(tileModel, row, column, 0f);
                    }
                    else
                    {
                        if ((cursorPosition.X == row) && (cursorPosition.Y == column))
                        {
                            if (board[row, column] == BoardColors.Empty)
                            {
                                DrawModel(highlightTileModel, row, column, 0f);
                            }
                            DrawModel(board.CurrentColor == BoardColors.White ?
                                whitePieceModel : blackPieceModel, row, column,
                                startingPieceHeight);
                        }
                        else
                        {
                            DrawModel(tileModel, row, column, 0f);
                        }
                        if (board[row, column] == BoardColors.Black)
                        {
                            DrawModel(blackPieceTileModel, row, column, 0f);
                        }
                        else if (board[row, column] == BoardColors.White)
                        {
                            DrawModel(whitePieceTileModel, row, column, 0f);
                        }
                    }
                }
            }
        }


        /// <summary>
        /// Draw the given model at the appropriate board position.
        /// </summary>
        /// <param name="model">The model to draw.</param>
        /// <param name="row">The board row.</param>
        /// <param name="column">The board column.</param>
        /// <param name="height">The height above the board.</param>
        /// <remarks>
        /// We know that none of these models have nontrivial transforms, 
        /// so we can skip the bone transform code.
        /// </remarks>
        private void DrawModel(Model model, int row, int column, float height)
        {
            // if the model is null, fail silently - given nothing to draw
            if (model == null)
            {
                return;
            }

            Matrix worldMatrix = Matrix.CreateTranslation(boardPosition +
                new Vector3(row * tileSize, 1f + height, column * tileSize));
            foreach (ModelMesh modelMesh in model.Meshes)
            {
                foreach (BasicEffect basicEffect in modelMesh.Effects)
                {
                    basicEffect.EnableDefaultLighting();
                    basicEffect.View = SimpleArcCamera.ViewMatrix;
                    basicEffect.Projection = SimpleArcCamera.ProjectionMatrix;
                    basicEffect.World = worldMatrix;
                }
                modelMesh.Draw();
            }
        }


        #endregion
    }
}
